### PR TITLE
Add/retry failed conversions

### LINKED ISSUE #28: Retry Failed Conversions
**Is your feature request related to a problem? Please describe.**
If a post ends up not getting converted properly, for example if the backend/server becomes nonresponsive, we'd need to have a way of retrying those conversions; they might go through the second time, which is sometimes that indeed does happen.

**Describe the solution you'd like**
There should be some kind of info on the WP-admin Run Conversion page telling the user whether some posts weren't converted successfully (giving a total number of failed posts, and possibly their IDs as well), with an option to retry converting only these failed ones.

### PR DESCRIPTION
**Pull request by iuravic**
Fixes #28 

**This PR brings two new features:**
- retrying conversion of just the posts which previously failed to get converted successfully,
- resetting an ongoing conversion, so that it can be completely restarted.

**TL;DR scenario** -- we'll create 300 test posts, that's 3 full batches, and rig the first half to fail conversion during the first run. Then we'll use the new functionality to retry converting just these previously failed Posts, and run in on those again. And we'll unrig them then, letting them pass this second time (this emulates a legitimate reason for failure now being cleared, such as backend being unresponsive before, but not any longer, etc). And we'll also test the resetting of an ongoing conversion, so that the conversion can be run all over again even though it was initialized and marked as "in progress".

**To test the PR:**
- generate test content:
    - first create exactly 300 public HTML Posts, and delete any other public Pages if there are any -- you can use the FakerPress Plugin to generate the 300 public HTML Posts
- install the plugin:
    - if the Plugin's already installed, first Deactivate & Delete it, then pull this branch and build it with `nvm use 10.10.0 && npm ci && npm run clean && npm run build:webpack`
    - additionally, if you're using a symlink to the Plugin's folder in your local env, and it may not be convenient for you to have to recreate the symlink after fully Deleting the plugin, feel free to pull, build and Activate according to these instructions https://github.com/Automattic/newspack-content-converter/pull/33#pullrequestreview-337412481 -- follow steps 1. and 2.
- purposely rig the first 150 Posts to fail during conversion (so that later on we have some failed ones to retry conversion):
    - run this SQL to set their content to invalid (one blank space causes a failure): `update ncc_wp_posts set post_content = '&nbsp;' order by ID limit 150;`
- run conversion for all content:
    - in WP-admin, go to "Newspack Content Converter" > "Run Conversion" > and click the big blue button "Start Conversion Now"
    - while conversion is running on the `/wp-admin/post-new.php?newspack-content-converter` page, open the Console, and observe that the first 150 posts fail with errors (you'll see output saying "A conversion error occured", and an error object with error message "No resulting blocks content") -- the whole first batch and a half of the second batch will fail, out of three batches total
    - after the 3 batches have completed with half of them failed, and half converted correctly, notice that the end message now says: "Warning: certain posts were not converted successfully..."
    - there click the "Back to Run Conversion page" link to `/wp-admin/admin.php?page=ncc-conversion` -- and then notice the new info section "Retry Converting Failed Posts" with another blue button "Retry Failed" -- don't click it yet
- now restore the content for the 150 failed posts, so that they will go through successfully the second time:
    - `update ncc_wp_posts nwp join wp_posts wp on wp.ID = nwp.ID set nwp.post_content = wp.post_content;`
- retry conversion only for the failed posts:
    - on the "Run Conversion" Page, click the "Retry Failed" second blue button
    - while conversion will now run 2 batches, in the Console observe that the 150 posts now get converted properly

If all this worked, it means that the new feature "retrying-conversion-of-only-failed-posts" works well :thumb_up.

**Let's also test the second feature added -- resetting an ongoing conversion:**
- start a conversion and purposely interrupt it:
    - in WP-admin, go to "Newspack Content Converter" > "Run Conversion" > and click the big blue button "Start Conversion Now"
    - once the conversion starts running on the `/wp-admin/post-new.php?newspack-content-converter` page, purposely close the active tab, and interrupt the conversion (another PR will add an extra warning to the user here, but there isn't any right now)
- reset ongoing conversion:
    - go to "Newspack Content Converter" > "Run Conversion" again, and notice the "Reset Ongoing Conversion in Order to Start Again" part
    - click the blue button "Reset Conversion"
    - confirm that the page now shows the "Start Conversion" button which lets you restart the conversion all over again, even though it's been closed unexpectedly -- if you wish, click it and confirm that the full 3 batches get converter successfully once again

### CODE REVIEW

File: assets/src/conversion/index.js [hunk 1]
```diff
@@ -19,6 +24,9 @@ class Conversion extends Component {
 			conversionBatchSize: '',
 			queuedBatchesCsv: '',
 			maxBatch: '',
+			hasConvertedPosts: '',
+			hasFailedConversions: '',
```
Comment (adekbadek) [inline assets/src/conversion/index.js]: Similar as in the other comment, these two booleans are oddly initiated as strings

File: assets/src/conversion/index.js [hunk 2]
```diff
@@ -133,23 +190,46 @@ class Conversion extends Component {
...
+									</b>
+								</li>
+							</ul>
+
+							<input
+								type="submit"
+								className="large"
+								id="retry_failed_button"
+								value={ __( 'Retry Failed' ) }
+								onClick={ event => this.handleOnClickInitializeRetryFailed( event ) }
```
Comment (adekbadek) [inline assets/src/conversion/index.js]: `handleOnClickInitializeRetryFailed` does not use the event argument, so it could just be:
```suggestion
								onClick={ this.handleOnClickInitializeRetryFailed }
```
(same goes for `this.handleOnClickInitializeConversion` handling above)

File: assets/src/conversion/index.js [hunk 3]
```diff
@@ -58,22 +70,50 @@ class Conversion extends Component {
 		} );
 	};
 
+	handleOnClickInitializeRetryFailed = () => {
+		return fetchInitializeRetryFailedConversion().then( response => {
+			if ( ! response || ! response.result || 'queued' != response.result ) {
+				return new Promise( ( resolve, reject ) => resolve() );
```
Comment (adekbadek) [inline assets/src/conversion/index.js]: 1. I don't understand why a Promise is returned here (and two lines above), since it's just a click handler 🤔(same in `handleOnClickInitializeConversion` above)
2. tip: a Promise that resolves immediately can be just [`Promise.resolve()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve)

File: assets/src/conversion/index.js [hunk 4]
```diff
@@ -94,23 +134,40 @@ class Conversion extends Component {
...
-							{ __(
-								'note: keep an eye on your active browser tabs currently performing the conversion'
-							) }
-						</i>
+						{ __(
+							'In case that your active conversion browser tab has been closed by accident, or it has been interrupted and closed unexpectedly, you may reset the conversion status here, and start converting all over again. Note that this will enable you to restart the conversion, but any previous results may be lost.'
+						) }
 					</p>
+					<input
+						type="submit"
```
Comment (adekbadek) [inline assets/src/conversion/index.js]: Maybe it's a matter of personal preference, but I'd replace these inputs with `button` elements, since that's what they are - there's no `form` to submit here ;)

File: assets/src/conversion/index.js [hunk 5]
```diff
@@ -133,23 +190,46 @@ class Conversion extends Component {
...
+									<b>
+										<u>{ countFailedConverting }</u>
+									</b>
+								</li>
+							</ul>
+
+							<input
+								type="submit"
+								className="large"
+								id="retry_failed_button"
```
Comment (adekbadek) [inline assets/src/conversion/index.js]: The ids on these inputs don't seem to be used anywhere - shouldn't they be removed?

File: lib/class-convertercontroller.php [hunk 6]
```diff
@@ -210,23 +268,68 @@ public function conversion_initialize() {
...
+
 		return rest_ensure_response(
 			[
-				'ids'       => $this->conversion_processor->set_next_conversion_batch_to_queue(),
-				'thisBatch' => max( $this->conversion_processor->get_conversion_queued_batches() ),
-				'maxBatch'  => $this->conversion_processor->get_conversion_max_batch(),
+				'ids'                      => $this->conversion_processor->set_next_conversion_batch_to_queue(),
+				'thisBatch'                => max( $this->conversion_processor->get_conversion_queued_batches() ),
+				'maxBatch'                 => $this->conversion_processor->get_conversion_max_batch(),
+				'hasIncompleteConversions' => $has_incomplete_conversions,
```
Comment (adekbadek) [inline lib/class-convertercontroller.php]: After running the conversion for the first time, the `Warning: certain posts were not converted successfully.` message does not appear. On subsequent conversions (before repairing the data, so it's still inconvertible), the message appears. 

I've tried debugging it, but didn't get past this point here. My hunch is that `$has_incomplete_conversions` value here is the one that is erroneously `false`. Is that possible?

File: assets/src/content-converter/index.js [hunk 7]
```diff
@@ -18,20 +22,37 @@ class ContentConverter extends Component {
...
 	}
 
 	componentDidMount() {
-		return fetchConversionBatch()
+		const { retryFailedConversions } = this.state;
+
+		// Get a batch of regular conversions, or retry the failed ones.
+		const fetchBatchPromise = ! retryFailedConversions
+			? fetchConversionBatch
+			: fetchRetryFailedConversionsBatch;
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/content-converter/index.js]: When I was writing this, I stopped on this same line and thought about the same thing. I'll change it to a positive form now, for easier reading.

File: assets/src/content-converter/index.js [hunk 8]
```diff
@@ -18,20 +22,37 @@ class ContentConverter extends Component {
 
 		this.state = {
 			isActive: null,
+			retryFailedConversions: props.retryFailedConversions,
 			postIds: '',
 			thisBatch: '',
 			maxBatch: '',
+			hasIncompleteConversions: '',
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/content-converter/index.js]: Good catch, I looked into all of the `state` variables in all of the components and set those to proper values now.

File: assets/src/content-converter/index.js [hunk 9]
```diff
@@ -18,20 +22,37 @@ class ContentConverter extends Component {
...
+				if ( response ) {
+					const { ids: postIds, thisBatch, maxBatch, hasIncompleteConversions } = response;
+					this.setState( {
+						postIds,
+						thisBatch,
+						maxBatch,
+						hasIncompleteConversions,
+						isActive: true,
+					} );
+					if ( this.state.postIds ) {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/content-converter/index.js]: Thanks so much for sharing this!
Instead of running this conditional in the callback, I changed the conditional to check the Reponse param instead, because it turns out simpler.

File: assets/src/conversion/index.js (hunk 1 above)
Comment (iuravic) [reply assets/src/conversion/index.js]: Updated in the comment above.

File: assets/src/conversion/index.js (hunk 2 above)
Comment (iuravic) [reply assets/src/conversion/index.js]: Nice, thanks! Updated.

File: assets/src/conversion/index.js (hunk 3 above)
Comment (iuravic) [reply assets/src/conversion/index.js]: True, returning Promises isn't necessary in these two cases. Updated to return a simple void.

File: assets/src/conversion/index.js (hunk 4 above)
Comment (iuravic) [reply assets/src/conversion/index.js]: I'd take this advice here in normal cases, but Thomas will do a complete redesign of the UI very soon, so let's leave that up to him :)

File: assets/src/conversion/index.js (hunk 5 above)
Comment (iuravic) [reply assets/src/conversion/index.js]: Hmm, for some reason I thought all input elements needed to have explicitly set unique ids... But I see now that's not true (at least not anymore). Updated!

File: lib/class-convertercontroller.php (hunk 6 above)
Comment (iuravic) [reply lib/class-convertercontroller.php]: Ah, awesome catch! I did a lot of code reshuffling and lost track of the sequence of execution which sets the state. Updated now, should work fine.
